_posts/wasmer-ruby-embedding-1.0.md

```diff
@@ -0,0 +1,458 @@
+# `ImportObject`.
+import_object = Wasmer::ImportObject.new
+
+# Let's write the Ruby function that is going to be imported.
+def sum(x, y)
+  x + y
+end
+
+sum_host_function = Wasmer::Function.new(
+  store,
+
+  # The host function `sum` as a Ruby symbol.
+  method(:sum),
+
+  # The host function signature:
+  #                         x                  y                    result
+  Wasmer::FunctionType.new([Wasmer::Type::I32, Wasmer::Type::I32], [Wasmer::Type::I32])
+)
+
+# Now let's register the `sum` import inside the `env` namespace.
+import_object.register(
+  "env",
+  {
+    :sum => sum_host_function,
+  }
+)
+
+# Let's instantiate the module!
+instance = Wasmer::Instance.new module_, import_object
+
+# And finally, call the `add_one` exported function!
+assert { instance.exports.add_one.(41) == 42 }
+```
+
+Specifying the function type is mandatory because we cannot infer the
+type of the function at runtime. Thus, it's up to the user to ensure
+that the types match. Let's see how we could define
+`sum_host_function` with `Proc`:
+
+```ruby
+sum_host_function = Wasmer::Function.new(
+  …,
+  # The host function `sum` as a Ruby proc.
+  Proc.new { |x, y| x + y },
+  …
+)
+```
+
+Finally, as a lambda:
+
+```ruby
+sum_host_function = Wasmer::Function.new(
+  …,
+  # The host function `sum` as a Ruby lambda function.
+  -> (x, y) { x + y },
+  …
+)
+```
+
+### Enumerable memories
+
+The `Memory` type represents a memory (yes, we are good at
+naming). It's a linear memory, that simply consists of a vector of
+bytes, and an optional maximum size. A memory created by the host or
+in WebAssembly code will be accessible and mutable from both host and
+WebAssembly.
+
+To make the life on the host side easier, we provide what we call
+“views”. For example, the `UInt32View` is a view over the memory data
+where elements are of kind `uint32`. We provide the following views:
+`Uint32View`, `Int32View`, `Uint32View`, `Int32View`, `Uint32View`,
+and `Int32View`. All these views include [the `Enumerable`
+mixin](https://ruby-doc.org/core-3.0.1/Enumerable.html). So it's now
+possible to do something like this:
+
+```ruby
+# Declare a WebAssembly module that:
+#
+# * exports a memory under the name `mem`,
+# * exports a function under the name `hello`, which returns
+#   a pointer to a string containing “Hello, World!”.
+wasm_bytes = Wasmer::wat2wasm(
+  (<<~WAST)
+  (module
+    (type $hello_t (func (result i32)))
+    (func $hello (type $hello_t) (result i32)
+      i32.const 42)
+    (memory $memory 1)
+    (export "hello" (func $hello))
+    (export "mem" (memory $memory))
+    (data (i32.const 42) "Hello, World!"))
+  WAST
+)
+
+# Create a store, compile the module, instantiate the module.
+store = Wasmer::Store.new
+module_ = Wasmer::Module.new store, wasm_bytes
+instance = Wasmer::Instance.new module_, nil
+
+# OK, here go. First, let's call `hello`. It returns a
+# pointer to the string in memory.
+pointer = instance.exports.hello.()
```

Comment:
Probably is:
```suggestion
pointer = instance.exports.hello()
```

Reply:
That's not a typo. That's… how Ruby works to call an object that acts like a function. The alternatives are `hello[]`, or `hello.call` or `f.===` (I swear I'm not joking!).

`hello.()` is a common form as far as I know.